### PR TITLE
debezium/dbz#2531 Resolve ambiguous partial rollbacks to the most recent transaction and prune stale deferred entries

### LINKED ISSUE #2531: Oracle: ambiguous partial rollback leaves stale deferred transaction that pins the offset SCN for the full deferred retention
## Bug report

**What Debezium connector do you use and what version?**

Oracle connector (LogMiner, buffered adapter), Debezium Server 3.6.1.Final. The affected code is unchanged on current `main` (`BufferedLogMinerStreamingChangeEventSource.java:637-680`), so the report applies to the latest version as well.

---

**What is the connector configuration?**

Relevant subset (Debezium Server property names, `debezium.source.` prefix stripped):

```
database.connection.adapter=logminer
log.mining.strategy=hybrid
log.mining.query.filter.mode=in
log.mining.buffer.type=memory (default)
log.mining.buffer.deferred.transaction.start=true
log.mining.buffer.deferred.transaction.retention.ms=<unset, default 24h>
log.mining.transaction.retention.ms=7200000
snapshot.mode=no_data
heartbeat.interval.ms=30000
```

---

**What is the captured database version and mode of deployment?**

Oracle 19c on Exadata, 2-node RAC, on-premises. High-volume OLTP (~0.8 TB redo/day). Connector runs in Kubernetes (on-premises), sink is Kafka.

---

**What behavior do you expect?**

When a transaction that only ever produced a START marker (a *deferred* transaction under `log.mining.buffer.deferred.transaction.start=true`) is rolled back, its deferred metadata entry is removed, so the offset low watermark (`offsetScn = min(oldestDeferredScn, minCacheScn) - 1`) can advance past it. A connector restart then resumes near the head of the redo stream.

---

**What behavior do you see?**

A rollback whose transaction id could not be fully resolved (LogMiner `ffffffff` sequence suffix, handled via `NO_SEQUENCE_TRX_ID_SUFFIX`) matched **more than one** known transaction by its usn+slot prefix. The ambiguous branch removes nothing:

```java
// BufferedLogMinerStreamingChangeEventSource.java (main, ~line 671)
// TODO: Investigate whether this scenario is possible and if so, how to disambiguate
LOGGER.warn("Unable to match partial transaction '{}' to a single transaction. ...");
```

The rolled-back transaction's deferred entry therefore stayed in the deferred map. Because deferred entries are retained for `log.mining.buffer.deferred.transaction.retention.ms` (default **24h**) and `log.mining.transaction.retention.ms` does not apply to them, the stale entry pinned the offset SCN for ~10 hours (a transaction STARTed at 06:40:23 pinned the offset until a 16:47 deploy restart):

```
16:40:23 WARN Offset SCN 13825932674208 has not changed in 25 mining session iterations. This may
         indicate long running transaction(s), active transactions: [31000c008d648101, 0c001f00e28dc303,
         560021003f794701]. Commit SCNs {1=13826203285704, 2=13826203285701}.
```

Two aggravating effects:

1. **The pin is invisible.** In deferred mode the mining window slides (`endScn-1`), so lag stays at zero and only the offset SCN silently stays behind. The stale-offset warning above prints only the *transaction cache* ids (`getActiveTransactionIds`); the deferred entry actually holding the watermark is never shown — the ids in the warning were rotating, second-lived OLTP transactions.
2. **Every restart re-mines the full distance.** On restart the connector resumed from the pinned offset SCN and had to re-mine ~10 hours of redo (hundreds of GB on this database) before producing anything new.

Notably, the re-mine itself *released* the entry: after the restart, START and rollback landed in the same mining session, the XID resolved fully, and the offset advanced (`Offsets as shutdown: scn=13825932674208` on the pre-restart pod vs. `scn=13825971349543` 24 minutes into the re-mine). Live, split across sessions, the same records leaked the entry. `gv$transaction` confirmed no long-running transaction actually existed at that time.

---

**Do you see the same behaviour using the latest released Debezium version?**

The ambiguous-match branch and the deferred map handling are identical on current `main` (verified by inspection; the only later change to this file was debezium/dbz#2151). We hit it on 3.6.1.Final in production.

---

**Do you have the connector logs, ideally from start till finish?**

Yes — INFO/WARN logs for the whole day are retained (Graylog). Excerpts above; happy to provide more (including the full sequence of `Offset SCN ... has not changed` warnings from 12:42 to 16:40 with rotating cache transaction ids, and the shutdown offset lines).

---

**How to reproduce the issue using our [tutorial](https://github.com/debezium/debezium-examples/tree/main/tutorial) deployment?**

Not practical via the tutorial: it requires an undo-slot (usn+slot) reuse where a transaction's rollback surfaces in a later mining session with an unresolvable sequence while another same-prefix transaction is known to the connector. It can be reproduced deterministically in a unit test against `BufferedLogMinerStreamingChangeEventSource` (deferred mode): register two deferred transactions sharing the usn+slot prefix, then deliver a `ROLLBACK` row whose transaction id carries the `ffffffff` suffix — the entry is never removed and `getOldestDeferredTransactionStartScn()` keeps returning the stale SCN. I can contribute this test together with a fix.

**Proposed fix** (PR to follow if the approach sounds right):

- Undo slot reuse is strictly ordered — of all same-prefix candidates, only the one with the **highest start SCN** can still be active, so an ambiguous partial rollback can safely be resolved to that candidate instead of being dropped.
- Older same-prefix *deferred* candidates are provably already terminated (their slot was reused) and can be pruned at the same time. Removing a deferred entry is lossless: if a DML for it ever arrives, `enqueueEvent` recreates the transaction from the DML row itself.
- Optionally, include the oldest deferred transaction id in the "Offset SCN ... has not changed" warning so the actual holder is visible to operators.

### PR DESCRIPTION
**Pull request by denizozogularas**
Fixes debezium/dbz#2531

## Description

The `ffffffff` partial-transaction fallback currently exists only for rollbacks, and its ambiguous branch removes nothing. Both gaps leak deferred (START-only) entries, which then pin the offset SCN low watermark for the full `log.mining.buffer.deferred.transaction.retention.ms` (default 24h) — invisibly, since in deferred mode the mining window keeps sliding and the stale-offset warning only prints transaction cache ids. In our production case (details in the issue) a restart then had to re-mine ~10 hours of redo.

Two changes, one per commit:

**1. Ambiguous partial rollbacks are resolved instead of dropped.** Oracle reuses an undo slot only after the transaction that previously occupied it has ended, so of all same-prefix candidates only the one with the highest start SCN can still be in progress — the rollback is applied to that candidate (same semantics as the existing single-match branch). Every older *deferred* candidate's slot has provably been reused, so those stale entries are pruned; older *cached* candidates are left untouched since they may carry events.

**2. Partial commits now clean up deferred entries too.** `handleCommitEvent` had no prefix fallback at all: a COMMIT arriving from a later mining session with an unresolved sequence missed the deferred map on the exact key and logged only at DEBUG, leaking the entry silently (this is the path our production case actually hit — the day's WARN logs contain no ambiguous-match warning). A partial commit now removes matching deferred entries by prefix, with a WARN. Cached transactions are deliberately not resolved by prefix on the commit side, because committing a guessed transaction would emit its events.

Removing a deferred entry is lossless: it holds no events, and if a DML for it ever arrived, `enqueueEvent` recreates the transaction from the DML row.

Unit tests cover both terminal event types: multiple deferred matches, cached-latest + stale-deferred, deferred-latest + older cached (untouched), and the three partial-commit shapes.

## PR Checklist
- [x] I have read the [contribution guidelines](https://github.com/debezium/debezium/blob/main/CONTRIBUTING.md) and the [governance document](https://github.com/debezium/governance/blob/main/GOVERNANCE.md) on PR expectations.
- [x] Minimal changes to code not directly related to your change (e.g. no unnecessary formatting changes or refactoring to existing code)
- [x] One feature/change per PR unless tightly coupled
- [x] Do a rebase on upstream `main`

### CODE REVIEW

**Comment by Naros** [comment]
Thanks for the PR, @denizozogularas.

I am a bit concerned because I think we have some overlapping concerns with #7472. In #7472 it relocates the undo handling inside `enqueueEvent`, when a rollback-flagged DML has no sequence (e.g. ends with `ffffffff`) and the prefix matches multiple cached transactions. 

This PR, paired with #7472, lands the codebase in an inconsistent state.

@kinolaev I think I'll merge this first since it's localized and much smaller; then we'll need to incorporate it into your PR and see whether adopting max-start-SCN resolution in the `enqueueEvent` branch works better than the warn-and-return approach.

I'd love feedback from you both on whether you concur with that approach.

**Comment by denizozogularas** [comment]
Thanks @Naros — merging this first and folding it into #7472 afterwards sounds right to me.

On adopting max-start-SCN resolution inside the `enqueueEvent` undo branch: I believe the same invariant carries over cleanly. An undo slot is reused only after the transaction occupying it has ended, so among same-prefix candidates only the one with the highest start SCN can still be in progress — that holds whether the reference comes from a ROLLBACK marker or a rollback-flagged DML. Warn-and-return leaves the referenced transaction unresolved, and at least in deferred mode that has a real cost: the stale entry pins the offset SCN low watermark for the deferred retention.

A field note that may help #7472's shape: we hit *both* silent variants in production on the same day — a deferred entry leaked by an unresolvable terminal event pinned the offset for ~10h (a restart then re-mined the whole distance), and two short-lived cached transactions whose COMMITs arrived cross-session as `ffffffff` were never finalized and were abandoned by `log.mining.transaction.retention.ms` two hours later, dropping their 7 events. The second commit here only covers the deferred side of partial COMMITs; resolving a *single-match cached* commit by prefix would be the remaining gap — I left it out since committing a guessed transaction emits events, but happy to follow whatever direction you prefer.

On the testing-farm failures: the buffered unit suite passes locally (20/20 in `DeferredMemoryStreamingChangeEventSourceTest`) — do those look related to this change or environmental? Glad to dig in if it's the former.

**Comment by kinolaev** [comment]
@denizozogularas thanks for the PR! Correct me if I'm wrong, but what I understood from https://github.com/debezium/dbz/issues/2531#issuecomment-5480560844 is that you never saw multiple transactions matched by the same prefix (which we assume is either impossible or a sign of corrupted state). Commit events with the `ffffffff` suffix were always the root cause, right?

If that's true, I don't think we need to do anything with rollbacks (full or partial) - we should instead handle the `ffffffff` suffix in `handleCommitEvent` the same way, including applying such commit events to cached transactions.

This brings me to another idea. @Naros, has anyone ever observed the `Cannot determine which transaction to rollback. Manual investigation required` warning? If not, why don't we always identify transactions by slot instead of by full XID? We could add a warning when we encounter a start event with a slot that's still in the cache or deferred transactions map, plus a configuration option to decide what to do with the old transaction in that case - commit or rollback. Additionally, we could store the sequence part in the transaction object, compare it with the incoming event's transaction sequence, and apply the same logic in case of a mismatch. I believe it would be dead code just to be on the safe side, because as far as I know, a slot is a physical address, so it's physically impossible to have two transactions with the same address at any given moment.

Wouldn't that be a simpler, more universal solution? @denizozogularas @Naros, WDYT?

About the merge order - I'm fine with merging this one first, I'll rebase and apply all the changes needed.

**Comment by denizozogularas** [comment]
@kinolaev Correct, that matches our evidence. We never observed the multi-match warning in the field (the day's WARN logs had neither "Unable to match" nor "Matched partial transaction"); the demonstrated root cause was a COMMIT arriving with the `ffffffff` suffix and finding no fallback at all. The multi-match handling in the first commit was reasoned defense rather than observation: once entries leak by any path, same-prefix collisions become possible but with the commit hole closed they shouldn't accumulate in the first place.

Agreed on handling `ffffffff` symmetrically in `handleCommitEvent`, including applying a single-match commit to a cached transaction with slot exclusivity that match isn't really a guess, and it would have *recovered* the 7 events we lost instead of dropping them. Two details I'd keep from this PR's shape: resolve to the most recent same-prefix candidate (slot reuse is strictly ordered), and for any *older* cached candidates prefer abandoning over emitting if one exists it is by definition stale state, and emitting stale events seems worse than dropping them.

On slot-based identity: conceptually agreed a slot is physically exclusive, so it is the natural key. Two wrinkles worth checking before going that far: the persisted buffer types (Infinispan/Ehcache) are keyed by full XID today, so restarts need a migration story; and keeping the sequence for your mismatch-detection warning is still worthwhile. Your incremental version warn on a START for an occupied slot plus a configurable disposition for the old entry , already delivers most of the diagnostic value cheaply.

Happy to validate any of this in our environment, we run a 3.6.1 fork with these two commits in production against a ~0.8 TB redo/day two-node RAC, so we can soak follow-up changes and report back.

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/buffered/BufferedLogMinerStreamingChangeEventSource.java [hunk 1]
```diff
@@ -667,13 +679,48 @@ else if (matchingTransactions.size() == 1) {
                     }
                 }
                 else {
```
Comment (Naros) [inline debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/buffered/BufferedLogMinerStreamingChangeEventSource.java:663]: This entire `else` and the `.size() == 1` branches can be folded into one another.

File: debezium-connector-oracle/src/test/java/io/debezium/connector/oracle/logminer/buffered/DeferredMemoryStreamingChangeEventSourceTest.java [hunk 2]
```diff
@@ -212,6 +212,111 @@ public void testPartialRollbackMatchesDeferredTransactionByPrefix() throws Excep
         }
     }
 
+    @Test
+    public void testPartialRollbackWithMultipleDeferredMatchesRollsBackLatestAndPrunesStale() throws Exception {
```
Comment (Naros) [inline debezium-connector-oracle/src/test/java/io/debezium/connector/oracle/logminer/buffered/DeferredMemoryStreamingChangeEventSourceTest.java:218]: These new tests are missing the `@FixFor` annotations.

**Review by Naros — CHANGES_REQUESTED** [review]
Hi @denizozogularas the PR requires a rebase and I've identified two changes that are needed. I tried to push the changes to the PR myself, but the PR does not allow pushes to your fork.

**Comment by Naros** [comment]
Regarding these, see inline:

> If that's true, I don't think we need to do anything with rollbacks (full or partial) - we should instead handle the ffffffff suffix in handleCommitEvent the same way, including applying such commit events to cached transactions.

I came across this too, but I wasn't sure whether it made sense to handle it in this PR or separately. I leave that up to you @denizozogularas. The one thing I found unusual is that the loop inside `removeDeferredTransactionsByPrefix` removes all matching transactions rather than just the last one. In practice, I would expect only one, but I wonder whether dropping them all could cause data loss if there are more than 1.

> This brings me to another idea. @Naros, has anyone ever observed the Cannot determine which transaction to rollback. Manual investigation required warning? If not, why don't we always identify transactions by slot instead of by full XID? We could add a warning when we encounter a start event with a slot that's still in the cache or deferred transactions map, plus a configuration option to decide what to do with the old transaction in that case - commit or rollback.

So while you are correct that today, the `{usn, slot}` is unique at any given moment, shifting to using that as the basis for transaction management is a relatively invasive change, and not one I'd prefer we pursue now. 

And regarding configuration options, that's one thing I want to move away from as much as we can. Right now, the connector has a tremendous number of options by comparison to any other connector, and that often gives the impression that the connector is harder to work with and get the configuration "right".

**Comment by denizozogularas** [comment]
Rebased on `main` and both review items are in:

- the `.size() == 1` and multi-match branches are folded into one — the max-start-SCN pick covers the single-candidate case naturally, and the stale-entry pruning loop is a no-op there;
- the six new tests now carry `@FixFor("debezium/dbz#2531")`.

On removing *all* matching deferred entries rather than just the matched one: it can't lose data, for two reasons that compound. A deferred entry holds no events — it is only `{txId, startScn, changeTime, userName, thread, clientId}` — so removal discards no DML; and if a DML for a removed entry ever did arrive, `enqueueEvent` recreates the transaction from the DML row itself. Beyond that, two live same-prefix transactions are impossible (a slot is only reused after its previous occupant ended), so every non-matched candidate is by definition a leftover whose terminal event was already missed — exactly the entries that otherwise pin the offset low watermark until the deferred retention expires. That said, you're right that in practice the list should almost always hold a single entry; the pruning only does work when state has already leaked.

On extending the commit-side handling to cached transactions: I'd keep it out of this PR. Applying a commit to a prefix-matched cached transaction emits its events, which raises questions this PR otherwise doesn't have to answer (older cached candidates, interplay with #7472's relocation of the undo handling), and #7472 is already touching that ground — happy to help validate whatever shape it takes there.

(Sorry about the fork not accepting your pushes — it's an organization-owned fork, so GitHub doesn't offer the maintainer-edit option. I'll keep the turnaround quick instead.)

**Comment by kinolaev** [comment]
> So while you are correct that today, the `{usn, slot}` is unique at any given moment, shifting to using that as the basis for transaction management is a relatively invasive change, and not one I'd prefer we pursue now.
>
> And regarding configuration options, that's one thing I want to move away from as much as we can. Right now, the connector has a tremendous number of options

I opened a draft PR #7995 as an experiment to see how invasive this shift would be. The result looks fairly straightforward to me - how does it look to you, @Naros?

I agree regarding the number of options, so I threw an `IllegalStateException` instead of logging a warning. Both behaviors - emitting or removing previous transaction events - have a 50% chance of being wrong, so manual investigation would be needed in that scenario.

This was just a quick experiment; I haven't run the integration tests or accounted for edge cases yet. I won't push this forward if you don't think it's the right direction, but I wanted to share it. Let me know if you'd like me to keep exploring this approach.

**Comment by Naros** [comment]
Hi @kinolaev, thanks for taking a stab at that. I do think that, even if we decide not to proceed, having the PoC and seeing the advantages and disadvantages can inform future decisions. It may also show something we didn't expect as a side effect, including bugs or even a better way forward. So I definitely think it's worth it either way. 

That said, I looked at the PR and left a comment, and we can continue the discussion there.

**Review by Naros — APPROVED** [review]

**Comment by Naros** [comment]
Applied manually due to merge conflicts, thanks @denizozogularas :tada:
